@AddCSRFToken sets underlying request to null

### Environment
* Play! 2.5.3
* java version "1.8.0_72"
* OSX 10.10

### Issue
I have a strange behaviour when using `AddCSRFToken` annotation.

The following code snippet:
``` java
@AddCSRFToken
public Result ping() {
    DynamicForm form = this.formFactory.form().bindFromRequest();
    return ok(CSRF.getToken(request()).map(CSRF.Token::value).orElse("no token"));
}
```
throws a NPE on `bindFromRequest()` that comes from `Form#requestData(Http.Request request)` when it tries to read the request's body.

After some investigation it seems that when `AddCSRFTokenAction` is called it sets a new `Http.Context` with a new `RequestImpl` based on the current RequestHeader containing the CSRF tags and a null underlying `Request` (https://github.com/playframework/playframework/blob/master/framework/src/play-filters-helpers/src/main/java/play/filters/csrf/AddCSRFTokenAction.java#L63-L73).

Sure. Here is the complete exception stack trace:

```
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:593)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at scala.concurrent.java8.FuturesConvertersImpl$CF.apply(FutureConvertersImpl.scala:21)
	at scala.concurrent.java8.FuturesConvertersImpl$CF.apply(FutureConvertersImpl.scala:18)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at scala.concurrent.BatchingExecutor$Batch$$anonfun$run$1.processBatch$1(BatchingExecutor.scala:63)
	at scala.concurrent.BatchingExecutor$Batch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:78)
	at scala.concurrent.BatchingExecutor$Batch$$anonfun$run$1.apply(BatchingExecutor.scala:55)
	at scala.concurrent.BatchingExecutor$Batch$$anonfun$run$1.apply(BatchingExecutor.scala:55)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
	at scala.concurrent.BatchingExecutor$Batch.run(BatchingExecutor.scala:54)
	at scala.concurrent.Future$InternalCallbackExecutor$.unbatchedExecute(Future.scala:599)
	at scala.concurrent.BatchingExecutor$class.execute(BatchingExecutor.scala:106)
	at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:597)
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:248)
	at scala.concurrent.Promise$class.complete(Promise.scala:55)
	at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:153)
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:91)
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91)
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:90)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:39)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:405)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.NullPointerException: null
	at play.data.Form.requestData(Form.java:181)
	at play.data.DynamicForm.bindFromRequest(DynamicForm.java:76)
	at controllers.Test.ping(Test.java:103)
	at router.Routes$$anonfun$routes$1$$anonfun$applyOrElse$26$$anonfun$apply$26.apply(Routes.scala:723)
	at router.Routes$$anonfun$routes$1$$anonfun$applyOrElse$26$$anonfun$apply$26.apply(Routes.scala:723)
	at play.core.routing.HandlerInvokerFactory$$anon$4.resultCall(HandlerInvoker.scala:157)
	at play.core.routing.HandlerInvokerFactory$$anon$4.resultCall(HandlerInvoker.scala:156)
	at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$14$$anon$3$$anon$1.invocation(HandlerInvoker.scala:136)
	at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:73)
	at play.http.HttpRequestHandler$1.call(HttpRequestHandler.java:54)
	at play.filters.csrf.AddCSRFTokenAction.call(AddCSRFTokenAction.java:76)
	at play.core.j.JavaAction$$anonfun$7.apply(JavaAction.scala:108)
	at play.core.j.JavaAction$$anonfun$7.apply(JavaAction.scala:108)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at play.core.j.HttpExecutionContext$$anon$2.run(HttpExecutionContext.scala:56)
	at play.api.libs.iteratee.Execution$trampoline$.execute(Execution.scala:70)
	at play.core.j.HttpExecutionContext.execute(HttpExecutionContext.scala:48)
	at scala.concurrent.impl.Future$.apply(Future.scala:31)
	at scala.concurrent.Future$.apply(Future.scala:492)
	at play.core.j.JavaAction.apply(JavaAction.scala:108)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5$$anonfun$apply$6.apply(Action.scala:112)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5$$anonfun$apply$6.apply(Action.scala:112)
	at play.utils.Threads$.withContextClassLoader(Threads.scala:21)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5.apply(Action.scala:111)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5.apply(Action.scala:110)
	at scala.Option.map(Option.scala:146)
	at play.api.mvc.Action$$anonfun$apply$2.apply(Action.scala:110)
	at play.api.mvc.Action$$anonfun$apply$2.apply(Action.scala:103)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:249)
	... 13 common frames omitted
```

--- a/framework/src/play-filters-helpers/src/main/java/play/filters/csrf/AddCSRFTokenAction.java
+++ b/framework/src/play-filters-helpers/src/main/java/play/filters/csrf/AddCSRFTokenAction.java
@@ -5,17 +5,16 @@ package play.filters.csrf;
 
 import java.util.concurrent.CompletionStage;
 
+import javax.inject.Inject;
+
 import play.api.libs.crypto.CSRFTokenSigner;
-import play.api.mvc.RequestHeader;
 import play.api.mvc.Session;
 import play.mvc.Action;
 import play.mvc.Http;
 import play.mvc.Http.Request;
+import play.mvc.Http.RequestBody;
 import play.mvc.Http.RequestImpl;
 import play.mvc.Result;
-import scala.Option;
-
-import javax.inject.Inject;
 
 public class AddCSRFTokenAction extends Action<AddCSRFToken> {
 
@@ -35,7 +34,8 @@ public class AddCSRFTokenAction extends Action<AddCSRFToken> {
 
     @Override
     public CompletionStage<Result> call(Http.Context ctx) {
-        RequestHeader request = CSRFAction.tagRequestFromHeader(ctx._requestHeader(), config, tokenSigner);
+        play.api.mvc.Request<RequestBody> request =
+            CSRFAction.tagRequestFromHeader(ctx.request()._underlyingRequest(), config, tokenSigner);
 
         if (CSRFAction.getTokenToValidate(request, config, tokenSigner).isEmpty()) {
             // No token in header and we have to create one if not found, so create a new token
@@ -50,7 +50,7 @@ public class AddCSRFTokenAction extends Action<AddCSRFToken> {
 
             // Also add it to the response
             if (config.cookieName().isDefined()) {
-                Option<String> domain = Session.domain();
+                scala.Option<String> domain = Session.domain();
                 ctx.response().setCookie(config.cookieName().get(), newToken, null, Session.path(),
                         domain.isDefined() ? domain.get() : null, config.secureCookie(), config.httpOnlyCookie());
             } else {
@@ -58,7 +58,7 @@ public class AddCSRFTokenAction extends Action<AddCSRFToken> {
             }
         }
 
-        final RequestHeader newRequest = request;
+        final play.api.mvc.Request<RequestBody> newRequest = request;
         // Methods returning requests should return the tagged request
         Http.Context newCtx = new Http.WrappedContext(ctx) {
             @Override
@@ -67,7 +67,7 @@ public class AddCSRFTokenAction extends Action<AddCSRFToken> {
             }
 
             @Override
-            public RequestHeader _requestHeader() {
+            public play.api.mvc.RequestHeader _requestHeader() {
                 return newRequest;
             }
         };

--- a/framework/src/play-filters-helpers/src/main/java/play/filters/csrf/RequireCSRFCheckAction.java
+++ b/framework/src/play-filters-helpers/src/main/java/play/filters/csrf/RequireCSRFCheckAction.java
@@ -3,6 +3,11 @@
  */
 package play.filters.csrf;
 
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+
+import javax.inject.Inject;
+
 import play.api.libs.Crypto;
 import play.api.mvc.RequestHeader;
 import play.api.mvc.Session;
@@ -11,10 +16,6 @@ import play.mvc.Action;
 import play.mvc.Http;
 import play.mvc.Result;
 import scala.Option;
-
-import javax.inject.Inject;
-import java.util.Map;
-import java.util.concurrent.CompletionStage;
 
 public class RequireCSRFCheckAction extends Action<RequireCSRFCheck> {
 
@@ -94,6 +95,6 @@ public class RequireCSRFCheckAction extends Action<RequireCSRFCheck> {
         }
 
         CSRFErrorHandler handler = injector.instanceOf(configuration.error());
-        return handler.handle(new Http.RequestImpl(request), msg);
+        return handler.handle(new play.core.j.RequestHeaderImpl(request), msg);
     }
 }

--- a/framework/src/play-filters-helpers/src/test/scala/play/filters/csrf/JavaCSRFActionSpec.scala
+++ b/framework/src/play-filters-helpers/src/test/scala/play/filters/csrf/JavaCSRFActionSpec.scala
@@ -86,6 +86,7 @@ object JavaCSRFActionSpec extends CSRFCommonSpecs {
   class MyAction extends Controller {
     @AddCSRFToken
     def add(): Result = {
+      require(Controller.request()._underlyingRequest() != null) // Make sure request is set
       // Simulate a template that adds a CSRF token
       import play.core.j.PlayMagicForJava.requestHeader
       Results.ok(CSRF.getToken.get.value)

